src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp

```diff
@@ -51,7 +81,23 @@ void setPyException(const Exception &exception, const py::module_ &pyModule,
       return pyClass(exception.what());
     }
   }();
+
+  // Set the error indicator to our converted C++->Python exception.
   PyErr_SetObject(pyClass.ptr(), pyInstance.ptr());
+
+  // Set the __cause__ of our new exception if raising from a previous
+  // exception.
+  if (fromExcVal != nullptr) {
+    PyObject *excVal = nullptr;
+    PyObject *excType = nullptr;
+    PyObject *excTraceback = nullptr;
+    PyErr_Fetch(&excType, &excVal, &excTraceback);
+    PyErr_NormalizeException(&excType, &excVal, &excTraceback);
+    Py_INCREF(fromExcVal);  // Bump to own two references, stolen by SetCause and SetContext.
```

Comment:
To confirm the logic here, `PyErr_Fetch` adds a reference to each of type, val, and traceback. Type and traceback are decref'd in the lambda above. So val (returned from the lambda) has ref=1. You then increment it here, it's ref=2. And then `PyException_SetCause` and `PyException_SetContext` steal 2 references, so we're now back to val has none. 

But reading `PyErr_Restore`, says each value requires a valid reference, as it takes it away. https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Restore

So I've lost track of intentions on references here.

Reply:
I think you've mixed up `fromExcVal` and `excVal` - the exception values from two different exceptions. I.e. the `PyErr_Restore` below steals a reference from `excVal`, not `fromExcVal`.

This code is mostly a copy-paste from pybind11's `raise_from` implementation.

Reply:
I was originally looking at the `PyErr_Fetch` on line 61 in the lambda, and counting from there, which assigns to `fromExcVal`. Bit confusing when there are two `PyErr_Fetch`s in play.

But, if you're not seeing any reference leaks, then I could quite possibly have miscounted from just eyeballing.

Reply:
Is then the missing piece of the puzzle the `PyErr_SetObject` inbetween on line 86? This adds a second exception into the mix.

I.e. we execute `PyErr_SetObject`, then (if there was a `fromExcVal`) we immediately `Fetch` that new exception, in order to `SetCause` and `SetContext` on it to the previous exception (`fromExcVal`), before `Restore`ing back the new exception.

Reply:
I would've said no to the `PyErr_SetObject` as it doesn't seem to interact with `fromExcVal` returned from the earlier lambda.

As we discussed earlier, tracking references isn't the best tooled up in Python, so I don't know if there's a best route forward on here, other than mark it mentally as 'something to watch'.

Reply:
Resolved offline.
I *think* (it's been a few days) that I mistaook the `PyErr_Restore` on line 99 for also using `fromExcVal`, but it doesn't.

So ref count of that from the lambda is 1, it's added to here, and then taken away twice in SetCause and SetContext.

We're square with the universe again.